prettyprinter-ansi-terminal/bench/LargeOutput.hs

```diff
@@ -0,0 +1,157 @@
+    case T.uncons x of
+        Nothing -> pure x
+        Just (t,ext) -> pure (T.cons (toUpper t) ext)
+
+anCol :: Color -> Doc AnsiStyle -> Doc AnsiStyle
+anCol = annotate . color
+
+prettyProgram :: Program -> Doc AnsiStyle
+prettyProgram (Program binds) = annotate italicized $ prettyBinds binds
+
+prettyBinds :: Binds -> Doc AnsiStyle
+prettyBinds (Binds bs) = align (vsep (map prettyBinding (M.assocs bs)))
+  where
+    prettyBinding (var, lambda) = pretty var <+> anCol Red "=" <+> prettyLambdaForm lambda
+
+prettyLambdaForm :: LambdaForm -> Doc AnsiStyle
+prettyLambdaForm (LambdaForm free bound body) = prettyExp . (<+> anCol Blue "->") . prettyBound . prettyFree $ anCol Blue "\\"
+  where
+    prettyFree | null free = id
+               | otherwise = (<> anCol Blue lparen <> hsep (map pretty free) <> anCol Blue rparen)
+    prettyBound | null bound = id
+                | null free = (<> hsep (map pretty bound))
+                | otherwise = (<+> hsep (map pretty bound))
+    prettyExp = (<+> prettyExpr body)
+
+prettyExpr :: Expr -> Doc AnsiStyle
+prettyExpr = \case
+    Let binds body ->
+        align (vsep [ anCol Red "let" <+> align (prettyBinds binds)
+                    , anCol Red "in" <+> prettyExpr body ])
+
+    Case scrutinee alts -> vsep
+        [ anCol Yellow "case" <+> prettyExpr scrutinee <+> anCol Yellow "of"
+        , indent 4 (align (vsep (map prettyAlt alts))) ]
+
+    AppF f [] -> annotate bold . anCol Green $ pretty f
+    AppF f args -> annotate bold . anCol Green $ pretty f <+> hsep (map pretty args)
+
+    AppC c [] -> annotate bold . anCol Green $ pretty c
+    AppC c args -> annotate bold . anCol Green $ pretty c <+> hsep (map pretty args)
+
+    AppP op x y -> annotate bold . anCol Green $ pretty op <+> pretty x <+> pretty y
+
+    LitE lit -> annotate bold . anCol Green $ pretty lit
+
+prettyAlt :: Alt -> Doc AnsiStyle
+prettyAlt (Alt con [] body) = pretty con <+> anCol Yellow "->" <+> prettyExpr body
+prettyAlt (Alt con args body) = pretty con <+> hsep (map pretty args) <+> anCol Yellow "->" <+> prettyExpr body
+
+scaled :: Gen a -> Gen a
+scaled = scale (\n -> n * 2 `quot` 3)
+
+fromTo :: Int -> Int -> Gen b -> Gen b
+fromTo a b gen = do
+    n <- choose (min a b, max a b)
+    resize n gen
+
+randomProgram
+    :: Int -- ^ Seed
+    -> Int -- ^ Generator size
+    -> Program
+randomProgram seed size = let MkGen gen = arbitrary in gen (mkQCGen seed) size
+
+main :: IO ()
+main = do
+    let prog = randomProgram 1 60
+        renderedProg = (renderLazy . layoutPretty defaultLayoutOptions { layoutPageWidth = Unbounded } . prettyProgram) prog
+        (progLines, progWidth) = let l = TL.lines renderedProg in (length l, maximum (map TL.length l))
+    putDoc ("Program size:" <+> pretty progLines <+> "lines, maximum width:" <+> pretty progWidth)
+
+    let render :: (SimpleDocStream AnsiStyle -> TL.Text) -> Program -> TL.Text
+        render r = r . layoutPretty defaultLayoutOptions . prettyProgram
```

Comment:
Let's better use the `Unbounded` `PageWidth` – that should be much cheaper.

Reply:
Also, why benchmark the layouting? Better limit it to the rendering.

Reply:
Ok, will fix

Reply:
> Also, why benchmark the layouting? Better limit it to the rendering, no?

I think we'd an `instance NFData (SimpleDocStream AnsiStyle)`, which we can't auto-derive since we can't provide an orphan `Generic` for `AnsiStyle`, as it's opaque.

Anyway, it's done this way in the `prettyprinter` benchmark this was based on, perhaps for the same reason.

Reply:
I thought `NFData` was only needed for the output, which is `Text`?!

Adding `NFData` and/or `Generic` instances sounds like a good idea though. Feel free to make an issue!

Reply:
Perhaps I'm misunderstanding how it works, but I thought that to accurately benchmark the rendering, we'd first have to force the input to that stage, which would be `SimpleDocStream`.